src/api/events/event-factory.js

```diff
@@ -0,0 +1,89 @@
+
+const DeviceStateChangedEvent = require('./device-state-changed-event');
+const ExecutionRegisteredEvent = require('./execution-registered-event');
+const ExecutionStateChangedEvent = require('./execution-state-changed-event');
+const DeviceStates = require('../device/device-states');
+const Command = require('../command');
+
+class EventFactory {
+    constructor(log) {
+        this.log = log;
+    }
+
+    createEvents(jsonEvent, deviceURLs) {
+        if (jsonEvent == null || !jsonEvent.name) {
+            this.log.warn(`JSON event is invalid: ${jsonEvent}.`)    
+            return null;
+        }
+
+        let events;
+        switch(jsonEvent.name) {
+            case 'DeviceStateChangedEvent':
+                events = this.createDeviceStateChangedEvent(jsonEvent);
+                break;
+            case 'ExecutionRegisteredEvent':
+                events = this.createExecutionRegisteredEvent(jsonEvent);
+                break;
+            case 'ExecutionStateChangedEvent':
+                events = this.createExecutionStateChangedEvent(jsonEvent, deviceURLs);
+                break;
+            default:
+                this.log.debug(`Ignore event ${jsonEvent.name}`);
+                events = [];
+        }
+        return events;
+    }
+
+    createDeviceStateChangedEvent(jsonEvent) {
+        let states = new DeviceStates(jsonEvent.deviceStates || []);
+        return [new DeviceStateChangedEvent(
+            this.log,
+            jsonEvent.timestamp,
+            jsonEvent.deviceURL,
+            states
+        )];
+    }
+
+    createExecutionRegisteredEvent(jsonEvent) {
+        if (jsonEvent.actions == null || jsonEvent.actions.length == 0) {
```

Comment:
We have `lodash` in the repository, the code could be changed to:
```suggestion
        if (isEmpty(jsonEvent.actions)) {
```